### DISCUSSION TITLE
Decoupling the VPU from the VGIC.

### DISCUSSION THREAD

**Discussion by pengzechen** [arceos-hypervisor/axvisor#12] [General]
There are three important structs in vgics.rs : (currently，only considering v2）VgicInt , VgicCpuPriv , Vgic.
 The latter two both contain VgicInt， it's defination as follows：
`struct VgicIntInnerMut {
    owner: Option<Vcpu>,
    in_lr: bool,
    lr: u16,
    enabled: bool,
    state: IrqState,
    prio: u8,
    targets: u8,
    cfg: u8,

    in_pend: bool,
    in_act: bool,
}`
I carefully checked every “owner” the whole file，they only use attr 'id', so why not use a "u32: id" or something else...

**Comment by pengzechen** [comment]
Also we need to decouple the gic from the VGIC.
The old implementation has always used static GICH，and It is in gic.rs
I don't want to import gic crate, My current thought is ，leave GICH (including it's init) in vigc, which kvm did.

**Reply by hky1999** [reply]
Theoretically speaking, GICH is only related to vGIC, I agree with you on this point.

**Comment by 915604903T** [comment]
As for the first question, I agree with you. We can discard the original `owner` field and use `vcpu_id` and `pcpu_id` to record the cpu information.

For GICH, I think your opinion is right.  We should remove the GICH implementation in arm_gic crate and implement it in the vgic crate instead.

**Comment by pengzechen** [comment]
There's another problem.
A vgic dev will operate real hardware(gicd)
The old implementation used "static GICD" (for example "GICD.setenable()", "GICD.setpriority()", etc...) do these directly, 
now we want to do these in a crate.
I think there are two potential solutions，
1）gic crate provide corresponding functions. like this `fn gic_setenable(int, en)`
     gic crate Hold the "static GICD" it self...
2）vgic save a pointer that can do these things, as a vigc init let the pointer point to GICD

I think Solution 1 is very likely to be reasonable. But we need change something in arceos ... 
If there is more reasonable method, I am all ears

**Reply by 915604903T** [reply]
I think the crate should not be related to hardware, whereas the 'static GICD' is strongly related to the hardware configuration. I am not sure where we need to expose an interface for GICD. Maybe in modules/axhal or somewhere else?
